### PR TITLE
Refactor types and constants

### PR DESCRIPTION
**Pull request by blazekadam**
Having types defined in objects such as `AbstractDataset` or `AbstractHook` complicates **cxflow** module structure and import dependencies.

In many cases, one had to import certain objects (and initialize the respective packages) just for the type hints. As a matter of fact, I found a cyclic dependency during my efforts to integrate `CatchSigint` hook to the `MainLoop`.

From now on, I suggest to 

- define all our types in `types.py` and include those types in the root package.
- define all `CONSTANTS` used on more than one place in `constants.py` (e.g., this PR moves `TRAIN_STREAM` from `MainLoop` to `constants.py` so that the `StopAfter` hook does not have to import `MainLoop` just for this constant)

### CODE REVIEW

File: cxflow/hooks/abstract_hook.py [hunk 1]
```diff
@@ -6,9 +6,10 @@
 import logging
 import inspect
 from typing import Iterable, Mapping
+from ..types import EpochData
 
-from ..datasets import AbstractDataset
-from ..utils.profile import Timer
+# from ..datasets import AbstractDataset
+# from ..utils.profile import Timer
```
Comment (FloopCZ) [inline cxflow/hooks/abstract_hook.py]: Remove those.

File: cxflow/hooks/abstract_hook.py [hunk 2]
```diff
@@ -67,7 +65,7 @@ def before_training(self) -> None:
         """
         pass
 
-    def after_batch(self, stream_name: str, batch_data: AbstractDataset.Batch) -> None:
+    def after_batch(self, stream_name: str, batch_data: object) -> None:
```
Comment (FloopCZ) [inline cxflow/hooks/abstract_hook.py]: Shouldn't there be `Batch` instead of `object`?

File: cxflow/hooks/abstract_hook.py [hunk 3]
```diff
@@ -91,7 +89,7 @@ def after_epoch(self, epoch_id: int, epoch_data: EpochData) -> None:
         """
         pass
 
-    def after_epoch_profile(self, epoch_id: int, profile: Timer.TimeProfile, extra_streams: Iterable[str]) -> None:
+    def after_epoch_profile(self, epoch_id: int, profile: object, extra_streams: Iterable[str]) -> None:
```
Comment (FloopCZ) [inline cxflow/hooks/abstract_hook.py]: ditto with `TimerProfile`

File: cxflow/hooks/show_progress.py [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 from . import AbstractHook
 from ..datasets import AbstractDataset
+from ..models import AbstractModel
```
Comment (FloopCZ) [inline cxflow/hooks/show_progress.py]: Never used?

File: cxflow/tests/main_loop_test.py [hunk 5]
```diff
@@ -246,7 +246,7 @@ def create_main_loop(self,  # pylint: disable=too-many-arguments
             model_class = TrainableModel
         model = model_class(dataset=dataset, log_dir=self.tmpdir,  # pylint: disable=redefined-variable-type
                             io={'in': ['input', 'target'], 'out': ['output']})
-        mainloop = MainLoop(model=model, dataset=dataset, hooks=hooks,
+        mainloop = cx.MainLoop(model=model, dataset=dataset, hooks=hooks,
                             skip_zeroth_epoch=skip_zeroth_epoch, **main_loop_kwargs)
```
Comment (FloopCZ) [inline cxflow/tests/main_loop_test.py]: Misaligned line.

File: cxflow/types.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+
+Batch = Mapping[str, Iterable[Any]]
+"""Batch type: :py:class:`typing.Mapping` of ``variable_name`` to an :py:class:`typing.Iterable` of examples."""
+
+Stream = Iterable[Batch]
+"""Stream type: :py:class:`typing.Iterable` of :py:attr:`Batch` es."""
+
+EpochData = Mapping[str, object]
+"""Epoch data type."""
```
Comment (FloopCZ) [inline cxflow/types.py]: We should add `TimeProfile` here.

File: cxflow/hooks/abstract_hook.py (hunk 2 above)
Comment (blazekadam) [reply cxflow/hooks/abstract_hook.py]: Yea... that is a relict from finding the import cycle.. sorry..

File: cxflow/hooks/abstract_hook.py (hunk 3 above)
Comment (blazekadam) [reply cxflow/hooks/abstract_hook.py]: Yea, `TimerProfile` should be in our `types.py` as well.

File: cxflow/types.py (hunk 6 above)
Comment (blazekadam) [reply cxflow/types.py]: Yep, I am on it.

Comment (blazekadam) [reply cxflow/types.py]: done

File: cxflow/types.py [hunk 7]
```diff
@@ -9,3 +9,6 @@
 
 EpochData = Mapping[str, object]
 """Epoch data type."""
+
+TimeProfile = Mapping[str, List[float]]
+"""Time profile type."""
```
Comment (FloopCZ) [inline cxflow/types.py]: No EOL at EOF.

File: cxflow/types.py [hunk 8]
```diff
@@ -9,3 +9,6 @@
 
 EpochData = Mapping[str, object]
 """Epoch data type."""
+
+TimeProfile = Mapping[str, List[float]]
```
Comment (FloopCZ) [inline cxflow/types.py]: List not imported.

**Pull request merged**